### ISSUE TITLE
Data block "Query returned no results" should return more information

### ISSUE THREAD

**Issue by cdobbyn** [hashicorp/terraform#27283]
### Current Terraform Version

We are in the process of updating to 0.13 but can't verify this just yet (large code base).
```
0.12.29
```

### Use-cases

If a data block fails to return results, the error should also indicate the resource that was trying to be collected. 

### Attempted Solutions

At present the only ways to do this are to turn down parallel work or through trace logging.

### Proposal

Return the resource that failed either above or below the error text:
```
Error: Your query returned no results. Please change your search criteria and try again
```

**Comment by apparentlymart** [comment]
Hi @cdobbyn,

Typically Terraform Core indicates the location of the relevant `resource`/`data` block if the underlying provider doesn't give a more specific item as part of its error message, but it seems like something is missing for failures during data resource reads where Terraform isn't applying that default.

However, the implementation of reading data sources has changed considerably in Terraform v0.14 both as part of some general refactoring and due to merging the refresh step into the plan step, so the behavior you've observed with Terraform v0.12 might not also occur in v0.14. I do see some code in v0.14 that seems to be [attaching source information to the returned diagnostics messages](https://github.com/hashicorp/terraform/blob/9ac8e3c55e01a2ee3465c9a2522cc0aa8662fb22/terraform/node_resource_abstract_instance.go#L1218), so perhaps this is an issue that we previously fixed in either a v0.13 or v0.14 release.

Since we're no longer developing Terraform v0.12 there won't be any changes to the behavior you've observed on that release, but we'll use this issue to prompt verifying whether the problem has been fixed in the meantime and if not then we can fix it in either a v0.14 point release or in the forthcoming v0.15.0, depending on the complexity of the change.

Thanks for reporting this!

**Comment by bflad** [comment]
Also potentially related: https://github.com/hashicorp/terraform-plugin-sdk/issues/561

**Comment by cdobbyn** [comment]
No problem @apparentlymart I'll see if I can reproduce the issue tomorrow evening in 0.13 and 0.14.

**Comment by cdobbyn** [comment]
To add some context (in case it helps), I originally saw this while using the aws provider (version 3.21.0) to lookup a aws_route_table. I can also try this another provider.

**Comment by marco-m-pix4d** [comment]
hello @apparentlymart 

a friendly reminder:

> but we'll use this issue to prompt verifying whether the problem has been fixed in the meantime and if not then we can fix it in either a v0.14 point release or in the forthcoming v0.15.0, depending on the complexity of the change.

This still happens with Terraform 0.14.x.

As @bflad already mentioned, this is related to hashicorp/terraform-plugin-sdk#561. See also hashicorp/terraform-provider-aws#15301.

My concern is that there are 3 projects involved: terraform-core, terraform-provider-aws, terraform-plugin-sdk and it might be not evident who does the first move.

What I can say is that this is impacting any refactoring of Terraform configuration, since one gets multiple
```
Error: Your query returned no results. Please change your search criteria and try again.
```
but not anymore any pointer to which terraform configuration file/line this happened, not even the name of the data source involved.

thanks!

**Comment by cdobbyn** [comment]
Thanks @marco-m-pix4d I had forgotten about this issue, thank you for reviving it.

**Comment by zswanson** [comment]
I can confirm this is an issue in terraform 0.13.x and 0.14.x. 

I've been meaning to open this exact bug/feature for some time. It's frustrating when we have multiple modules in a component that are all doing many lookups. We've tried using the terraform debug output to figure it out but usually just resort to having to comb through the modules to and check to all of them to see where we messed up a filter.

**Comment by marco-m-pix4d** [comment]
Ping @apparentlymart ?

**Comment by zswanson** [comment]
This can be reproduced in terraform 0.14.5 with the following sample using AWS provider.

```
terraform {}


provider "aws" {
  region = "us-east-1"
}


data "aws_security_group" "this" {

  filter {
    name   = "tag:ThisTag"
    values = ["ProbablyDoesNotExist"]
  }
}
```

> $ terraform plan
> Error: no matching SecurityGroup found

**Comment by danieldreier** [comment]
Thanks @zswanson - this is super helpful. I have tried this on 0.14.5 and can confirm exactly the same output as you showed. If I understand correctly, you would like to get output that indicates that a `aws_security_group` was not found, and the `no matching SecurityGroup found` output does not make it easy to find the source of this issue in the terraform code because "SecurityGroup" doesn't actually appear anywhere in the code that produced this error. Am I understanding the problem and desired behavior correctly?

**Comment by zswanson** [comment]
Ideally I'd like to see the full namespace of the offending data lookup. Such as 'module.foo.data.aws_security_group.bar', 
While the single security group illustrates the problem here, this sort of data error comes up with all sorts of other resources lookups. In a real composition we might have lots of data lookups of the same resource type, so we need more information on which one actually failed to return a result. I can test a few other data lookup types but I'm fairly certain I've encountered a few that didn't even tell us the object type, the error was more like 'filter returned no results' and that was it.

These are, admittedly, normally user driven errors where we put a typo in somewhere, or someone destroyed a resource elsewhere not realizing it was being used indirectly by another module, but the output right now makes it painful to diagnose the source of the problem.

**Comment by damon-atkins** [comment]
The larger the code base the longer it takes to find the issue.   If you lucky you have changed one value before getting this error and hence known what to fix.  Its a show stopper for easy of use of terraform.

**Comment by damon-atkins** [comment]
@danieldreier  is this on the roadmap to be fixed, both open source and commercial users would be effected.

**Comment by bflad** [comment]
There were two sides and fixes for this issue:

* Terraform CLI (this codebase) when receiving partially complete diagnostics from the provider could omit the configuration source in certain scenarios. The relevant fix for this was merged in for inclusion in Terraform CLI version 0.15.0-beta1: https://github.com/hashicorp/terraform/pull/27710
* Terraform Plugin SDK, the primary library used by many Terraform Providers today, could send partially complete diagnostics in certain scenarios. The relevant fix for this was merged in for inclusion in Terraform Plugin SDK version 2.4.4: https://github.com/hashicorp/terraform-plugin-sdk/pull/696

Terraform Providers depending on the Terraform Plugin SDK will need to upgrade to version 2.4.4 in their codebases to receive the fix. Once either Terraform CLI 0.15.0-beta1 (and later) is being used _or_ a released provider version includes the Terraform Plugin SDK update, this issue should be resolved.

For example, we will be updating the Terraform AWS Provider dependency as part of the version 3.30.0 release. Practitioners can use Terraform CLI 0.15.0-beta1 (and later) _or_ Terraform AWS Provider 3.30.0 when released to receive the fix.

I am going to leave this issue open here as the maintainers here can provide additional context or if there are any potential plans for backporting this to the Terraform CLI 0.14.x releases as well.

**Comment by pselle** [comment]
> I am going to leave this issue open here as the maintainers here can provide additional context or if there are any potential plans for backporting this to the Terraform CLI 0.14.x releases as well.

This has been backported and will be in the next 0.14 release!

**Comment by jbardin** [comment]
Closed via #27710
